packages/embark-compiler/src/index.ts

```diff
@@ -14,75 +16,166 @@ class Compiler {
     this.isCoverage = options.isCoverage;
 
     // embark.events.setCommandHandler("compiler:contracts", this.compile_contracts.bind(this));
-    embark.events.setCommandHandler("compiler:contracts:compile", this.compile_contracts.bind(this));
+    embark.events.setCommandHandler(
+      "compiler:contracts:compile",
+      this.compile_contracts.bind(this),
+    );
   }
 
-  private compile_contracts(contractFiles: any[], cb: any) {
+  private compile_contracts(contractFiles: any[], cb: Callback<any>) {
     if (contractFiles.length === 0) {
       return cb(null, {});
     }
 
-    const compiledObject: {[index: string]: any} = {};
+    async.waterfall(
+      [
+        (next: Callback<any[]>) => {
+          this.plugins.runActionsForEvent(
+            "compiler:contracts:compile:before",
+            contractFiles,
+            (err?: Error | null, files: any[] = []) => {
+              if (err) {
+                return next(err);
+              }
+              next(
+                null,
+                files.map((file: any) =>
```

Comment:
can we add something like

```
const filesToReturn = files.map(.....
next(null, filesToReturn)
```

Reply:
Sure, the linter (our present settings) just doesn't allow shadowing so I couldn't keep it named `contractFiles` even though it essentially is `contractFiles` in the outer scope.

Reply:
Renamed it to `contractFilesArr`.